Add mcp2515 driver

akif999 opened the pull request:
I added a driver for a CAN controller that works with a SPI interface called mcp2515.
This allows you to communicate with other microcontrollers using the CAN (Controller Aread Network) protocol.

Tested with the following boards.
* Adafruit Feather-m4
* SeeedStudio Seeeduino Xiao

akif999 committed:
mcp2515: add support for mcp2515 CAN device

deadprogram commented:
Hello @akif999 in order to pass the CI tests, you have to either add your driver to the list to skip the unit tests here: https://github.com/tinygo-org/drivers/blob/dev/Makefile#L182-L184 or else add a unit test such as https://github.com/tinygo-org/drivers/blob/dev/mcp23017/device_test.go

akif999 commented:
Hello @deadprogram . Thank you for your advice.  I added this package to `NOTESTS`. and I've rebased this branch on the latest `dev` branch.

deadprogram commented:
@sago35 any feedback on this driver?

@akif999 have you seen https://github.com/tinygo-org/tinygo/pull/1858 that was just merged into TinyGo `dev` branch?

sago35 commented:
I'll check it out now.

sago35 commented:
Please resolve the merge conflict. @akif999

akif999 commented:
> @akif999 have you seen tinygo-org/tinygo#1858 that was just merged into TinyGo dev branch?

@deadprogram  Yes, I've seen it.

@sago35  I merged conflict and rebased again.

sago35 commented on examples/mcp2515/main.go:
```diff
@@ -0,0 +1,55 @@
+	can.Configure()
+	err := can.Begin(mcp2515.CAN500kBps, mcp2515.Clock8MHz)
+	if err != nil {
+		panic(err)
```
You may change it to faileMessage.

akif999 replied:
updated.

sago35 commented on mcp2515/mcp2515.go:
```diff
@@ -0,0 +1,791 @@
+// Reset resets mcp2515.
+func (d *Device) Reset() error {
+	d.cs.Low()
+	defer d.cs.High()
+	_, err := d.spi.readWrite(mcpReset)
+	// time.Sleep(time.Microsecond * 4)
+	d.cs.High()
```
High() is called twice, so it is better to fix it. If possible, it is better to call d.cs.High() before time.Sleep(), so it may be better to delete the defer side.
High() in other functions as well, please check if they need to be modified.

akif999 replied:
updated.

sago35 commented:
@akif999 
Thank you for the contribution.
I think it's good for the most part, but I did have a few comments.

sago35 commented on README.md:
```diff
@@ -90,6 +90,7 @@ The following 61 devices are supported.
 | [MAX7219 & MAX7221 display driver](https://datasheets.maximintegrated.com/en/ds/MAX7219-MAX7221.pdf) | SPI |
 | [MCP23017 port expander](https://ww1.microchip.com/downloads/en/DeviceDoc/20001952C.pdf) | I2C |
 | [MCP3008 analog to digital converter (ADC)](http://ww1.microchip.com/downloads/en/DeviceDoc/21295d.pdf) | SPI |
+| [MCP2515 Stand-Alone CAN Controller with SPI Interface](https://ww1.microchip.com/downloads/en/DeviceDoc/MCP2515-Family-Data-Sheet-DS20001801K.pdf) | SPI |
```
Please update the numbers as the number of devices has increased.

```diff
-The following 61 devices are supported.
+The following 62 devices are supported.
```

deadprogram replied:
Actually now it will be 63, if this is able to be merged next.

deadprogram replied:
And now it would be number 64 `2^6` which is a very fortuitous number indeed.

akif999 replied:
updated.   I counted it and listed it as 64.

deadprogram commented:
I would rename `mcp2515_def.go` to just `registers.go` since it is already namespaced by package, and that is pretty typical for other drivers.

deadprogram commented:
@akif999 any chance you might be able to get these last few items addressed, and then squash this into a single commit? I would love to get it into the release which is going out today. :heart: 

Thank you!

akif999 commented:
@deadprogram  @sago35 
I fixed the code according to your comments. and I squashed and rebased this.

difference is here (showing in another branch).
https://github.com/akif999/drivers/commit/f1eb570cb172b5aa4cfeadaa9f60016c14f66564
https://github.com/akif999/drivers/commit/5355f30c70d714f02532d831ff31ce5c724420b7

Please check this. Thank you.

sago35 commented:
I'm going to check it out now.

sago35 commented on mcp2515/mcp2515.go:
```diff
@@ -0,0 +1,784 @@
+	if err != nil {
+		return err
+	}
+	d.cs.High()
```
There are still places where we are making double calls to defer.
If you have a reason to keep them, please write a comment in the source code.

sago35 replied:
Everything is OK except here.

akif999 replied:
updated.  I added comment in the source code.

sago35 reviewed (approved):
Looks good to me.
Thank you for contributions @akif999 .

sago35 commented:
The review is finished and it is ready to merge. @deadprogram

deadprogram commented:
Great work on this @akif999 and thank you very much @sago35 for all the help reviewing.

Now merging, thanks again!
